### PR TITLE
[serve] Run sync callables in `asyncio` executor to avoid blocking the loop

### PR DESCRIPTION
**Pull request by edoakes**
## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by edoakes** [comment]
Microbenchmark results with and without the change:
```
(ray) eoakes@Edwards-MacBook-Pro-2 serve % diff benchmarks/no_executor.txt benchmarks/in_executor.txt
2,3c2,3
<       single client small data 791.16 +- 45.89 requests/s
<       8 clients small data 1267.75 +- 7.54 requests/s
---
>       single client small data 866.99 +- 11.62 requests/s
>       8 clients small data 689.55 +- 204.17 requests/s
5,6c5,6
<       single client small data 899.14 +- 1.71 requests/s
<       8 clients small data 2816.31 +- 31.93 requests/s
---
>       single client small data 638.93 +- 93.66 requests/s
>       8 clients small data 2365.33 +- 171.19 requests/s
8,9c8,9
<       single client small data 839.42 +- 19.4 requests/s
<       8 clients small data 2702.35 +- 29.84 requests/s
---
>       single client small data 665.74 +- 98.41 requests/s
>       8 clients small data 2586.31 +- 58.95 requests/s
11,12c11,12
<       single client small data 739.12 +- 8.24 requests/s
<       8 clients small data 2616.61 +- 8.83 requests/s
---
>       single client small data 709.7 +- 6.4 requests/s
>       8 clients small data 2292.97 +- 154.68 requests/s
14,15c14,15
<       single client small data 752.73 +- 3.02 requests/s
<       8 clients small data 2683.5 +- 63.91 requests/s
---
>       single client small data 588.23 +- 38.24 requests/s
>       8 clients small data 2299.26 +- 100.08 requests/s
17,18c17,18
<       single client small data 707.58 +- 2.96 requests/s
<       8 clients small data 2662.97 +- 6.23 requests/s
---
>       single client small data 612.68 +- 72.04 requests/s
>       8 clients small data 2204.82 +- 93.43 requests/s
20,21c20,21
<       single client small data 863.78 +- 6.95 requests/s
<       8 clients small data 1269.88 +- 10.91 requests/s
---
>       single client small data 642.81 +- 49.43 requests/s
>       8 clients small data 1110.17 +- 25.55 requests/s
23,24c23,24
<       single client small data 881.01 +- 18.8 requests/s
<       8 clients small data 2796.18 +- 29.1 requests/s
---
>       single client small data 764.38 +- 85.44 requests/s
>       8 clients small data 2396.0 +- 146.61 requests/s
26,27c26,27
<       single client small data 691.03 +- 20.44 requests/s
<       8 clients small data 2590.25 +- 27.12 requests/s
---
>       single client small data 665.91 +- 7.4 requests/s
>       8 clients small data 2424.0 +- 30.6 requests/s
29,30c29,30
<       single client small data 734.39 +- 16.96 requests/s
<       8 clients small data 2683.58 +- 19.97 requests/s
---
>       single client small data 658.21 +- 5.14 requests/s
>       8 clients small data 2487.89 +- 28.53 requests/s
32,33c32,33
<       single client small data 417.12 +- 16.68 requests/s
<       8 clients small data 796.1 +- 2.9 requests/s
---
>       single client small data 367.45 +- 22.57 requests/s
>       8 clients small data 761.92 +- 21.41 requests/s
35,36c35,36
<       single client small data 442.52 +- 0.95 requests/s
<       8 clients small data 1353.99 +- 23.72 requests/s
---
>       single client small data 365.19 +- 34.66 requests/s
>       8 clients small data 1207.97 +- 53.52 requests/s
38,39c38,39
<       single client small data 420.69 +- 7.54 requests/s
<       8 clients small data 1340.0 +- 30.82 requests/s
---
>       single client small data 331.29 +- 37.18 requests/s
>       8 clients small data 1186.33 +- 31.21 requests/s
41,42c41,42
<       single client small data 389.52 +- 3.99 requests/s
<       8 clients small data 1255.39 +- 7.56 requests/s
---
>       single client small data 362.21 +- 21.32 requests/s
>       8 clients small data 1171.07 +- 35.11 requests/s
44,45c44,45
<       single client small data 396.82 +- 2.79 requests/s
<       8 clients small data 1291.72 +- 17.51 requests/s
---
>       single client small data 353.02 +- 13.74 requests/s
>       8 clients small data 1167.9 +- 18.31 requests/s
47,48c47,48
<       single client small data 376.96 +- 5.89 requests/s
<       8 clients small data 1259.12 +- 13.84 requests/s
---
>       single client small data 326.61 +- 26.67 requests/s
>       8 clients small data 1254.49 +- 7.12 requests/s
50,51c50,51
<       single client small data 401.18 +- 6.1 requests/s
<       8 clients small data 800.08 +- 5.17 requests/s
---
>       single client small data 332.77 +- 38.04 requests/s
>       8 clients small data 642.16 +- 43.57 requests/s
53,54c53,54
<       single client small data 329.81 +- 168.13 requests/s
<       8 clients small data 1166.04 +- 216.68 requests/s
---
>       single client small data 361.97 +- 14.76 requests/s
>       8 clients small data 1195.47 +- 44.25 requests/s
56,57c56,57
<       single client small data 404.97 +- 0.82 requests/s
<       8 clients small data 1278.49 +- 9.97 requests/s
---
>       single client small data 354.88 +- 6.75 requests/s
>       8 clients small data 1167.46 +- 24.11 requests/s
59,60c59,60
<       single client small data 290.86 +- 168.2 requests/s
<       8 clients small data 1047.7 +- 196.22 requests/s
---
>       single client small data 315.87 +- 15.01 requests/s
>       8 clients small data 1120.67 +- 15.62 requests/s
```

File: python/ray/serve/benchmarks/microbenchmark.py [hunk 1]
```diff
@@ -61,55 +59,32 @@ async def do_queries(self, num, data):
...
+        def __call__(self, req):
+            return b"ok"
 
     @serve.deployment(
-        name=deployment_name,
         num_replicas=num_replicas,
         max_concurrent_queries=max_concurrent_queries,
     )
-    class D:
+    class AsyncCallee:
```
Comment (alexeykudinkin) [inline python/ray/serve/benchmarks/microbenchmark.py:84]: What's the difference b/w Sync and Async targets here (besides batching)?

Should we consider making different targets mimicking diff workloads? For ex, CPU-bound, blocking I/O, etc

File: python/ray/serve/_private/replica.py [hunk 2]
```diff
@@ -432,25 +432,28 @@ async def invoke_single(self, request_item: Query) -> Tuple[Any, bool]:
...
+                len(args) == 1
+                and isinstance(args[0], starlette.requests.Request)
+                and len(inspect.signature(runner_method).parameters) == 0
+            ):
+                args = tuple()
+                kwargs = dict()
+
+            # If the callable is a sync method, run it in the asyncio loop's
+            # default executor.
+            if inspect.iscoroutinefunction(runner_method):
```
Comment (alexeykudinkin) [inline python/ray/serve/_private/replica.py:448]: Why would we want to differentiate b/w sync/async here? 

There could also be poorly implemented async method that might have CPU-bound prologue that would be blocking the event-loop before yielding it.

File: python/ray/serve/_private/replica.py [hunk 3]
```diff
@@ -432,25 +432,28 @@ async def invoke_single(self, request_item: Query) -> Tuple[Any, bool]:
...
-                    # The method doesn't take in anything, including the request
-                    # information, so we pass nothing into it
-                    result = await method_to_call()
-                else:
-                    # Will throw due to signature mismatch if user attempts to
-                    # call with non-empty args
-                    result = await method_to_call(*args, **kwargs)
+                loop = asyncio.get_running_loop()
+                coro = loop.run_in_executor(
+                    None, lambda: runner_method(*args, **kwargs)
```
Comment (alexeykudinkin) [inline python/ray/serve/_private/replica.py:453]: Do you think we should control target executor pool here as well? 

We can stick to the default for now, but in the future that might be a good idea to provide better isolation keeping all of the user-code to a standalone pool

**Review by alexeykudinkin — COMMENTED** [review]
LGTM! Minor comments

**Comment by edoakes** [comment]
Note to self: https://docs.python.org/3/library/asyncio-dev.html#debug-mode

**Comment by edoakes** [comment]
Planning to hold off and merge this after 2.5 branch cut given it's fairly high risk

**Comment by alexeykudinkin** [comment]
> Microbenchmark results with and without the change (no noticeable difference):

@edoakes can you share benchmark code you're using? I actually see quite a bit of variation in the numbers (op / s)

**Comment by edoakes** [comment]
@alexeykudinkin it's the microbenchmark I've updated in this PR at `python/ray/serve/benchmarks/microbenchmark.py`

**Comment by edoakes** [comment]
Note to self, need to handle:

- `__call__`
- `reconfigure`
- `@serve.batch`

**Comment by edoakes** [comment]
Not currently needed, using the existing Ray `concurrency_group` mechanism for the power of two choices routing instead

**Pull request closed without merging**